client_request_metrics_test.py

```diff
@@ -0,0 +1,784 @@
+from cassandra.query import SimpleStatement
+
+since = pytest.mark.since
+
+jmx = None
+
+KEYSPACE = 'ks'
+FAIL_WRITE_KEYSPACE = 'fail_keyspace'
+VIEW_KEYSPACE = 'view_keyspace'
+TABLE = 't'
+VIEW = 'mv'
+TOMBSTONE_FAILURE_THRESHOLD = 20
+TOMBSTONE_FAIL_KEY = 10000001
+
+
+class NoException(BaseException):
+    pass
+
+
+@since('4.0')
+class TestClientRequestMetrics(Tester):
+
+    @pytest.fixture(autouse=True)
+    def fixture_add_additional_log_patterns(self, fixture_dtest_setup):
+        fixture_dtest_setup.ignore_log_patterns = (
+            'Testing write failures',  # The error to simulate a write failure
+            'ERROR WRITE_FAILURE',  # Logged in DEBUG mode for write failures
+            f"Scanned over {TOMBSTONE_FAILURE_THRESHOLD+1} tombstones during query"  # Caused by the read failure tests
+        )
+
+    def setup_once(self):
+        cluster = self.cluster
+        cluster.set_configuration_options({'read_request_timeout_in_ms': 3000,
+                                           'write_request_timeout_in_ms': 3000,
+                                           'phi_convict_threshold': 12,
+                                           'tombstone_failure_threshold': TOMBSTONE_FAILURE_THRESHOLD,
+                                           'enable_materialized_views': 'true'})
+        cluster.populate(2, debug=True)
+        self.start_cluster()
+        node1 = cluster.nodelist()[0]
+
+        global jmx
+        jmx = JolokiaAgent(node1)
+        jmx.start()
+
+        s = self.session = self.patient_exclusive_cql_connection(node1, retry_policy=FallthroughRetryPolicy(), request_timeout=30)
+        for k in [KEYSPACE, FAIL_WRITE_KEYSPACE]:
+            s.execute(f"CREATE KEYSPACE {k} WITH replication = {{'class': 'SimpleStrategy', 'replication_factor': '2'}}")
+            s.execute(f"CREATE TABLE {k}.{TABLE} (k int, c int, v int, PRIMARY KEY (k,c))")
+
+        s.execute(f"CREATE KEYSPACE {VIEW_KEYSPACE} WITH replication = {{'class': 'SimpleStrategy', 'replication_factor': '1'}}")
+        s.execute(f"CREATE TABLE {VIEW_KEYSPACE}.{TABLE} (k int, c int, v int, PRIMARY KEY (k,c))")
+        s.execute(f"CREATE MATERIALIZED VIEW {VIEW_KEYSPACE}.{VIEW} AS SELECT * FROM {TABLE} WHERE c IS NOT NULL AND k IS NOT NULL PRIMARY KEY (c,k);")
+
+        for c in range(TOMBSTONE_FAILURE_THRESHOLD + 1):
+            self.session.execute(f"DELETE FROM {KEYSPACE}.{TABLE} WHERE k={TOMBSTONE_FAIL_KEY} AND c={c}")
+
+        node1.watch_log_for("Created default superuser role 'cassandra'")  # don't race with async default role creation, which creates a write
+        node1.watch_log_for('Completed submission of build tasks for any materialized views defined at startup', filename='debug.log')  # view builds cause background reads
+
+    def start_cluster(self):
+        self.cluster.start(jvm_args=[f"-Dcassandra.test.fail_writes_ks={FAIL_WRITE_KEYSPACE}"])
+
+    def test_client_request_metrics(self):
+        self.setup_once()
```

Comment:
I understand that the purpose of `setup_once` is preparing the scenario to run all the tests at once, so we don't spend so many resources initializing the cluster on each test, is this right?

Reply:
Accurate. I will add a comment.